lib/util/extractSourceMap.js

```diff
@@ -0,0 +1,296 @@
+		replacementString: match ? match[0] : null
+	};
+}
+
+function getAbsolutePath(fs, context, request, sourceRoot) {
+	if (sourceRoot) {
+		if (isAbsolute(sourceRoot)) {
+			return join(fs, sourceRoot, request);
+		}
+
+		return join(fs, join(fs, context, sourceRoot), request);
+	}
+
+	return join(fs, context, request);
+}
+
+function fetchFromDataURL(sourceURL, callback) {
+	const dataURL = /^data:(?:[^;,]+)?(?:;[^;,]+)*?(?:;(base64))?,(.*)$/i.exec(
+		sourceURL
+	);
+
+	if (dataURL) {
+		const encodingName = dataURL[1] ? "base64" : "ascii";
+		return callback(
+			null,
+			Buffer.from(dataURL[2], encodingName).toString("utf-8")
+		);
+	}
+
+	callback(
+		new Error(`Failed to parse source map from "data" URL: ${sourceURL}`)
+	);
+}
+
+function fetchFromFilesystem(fs, sourceURL, callback) {
+	fs.readFile(sourceURL, (err, bufferOrString) => {
+		if (err)
+			return callback(
+				new Error(`Failed to parse source map from '${sourceURL}' file: ${err}`)
+			);
+		callback(null, { path: sourceURL, data: bufferOrString.toString("utf-8") });
+	});
+}
+
+function fetchPathsFromFilesystem(
+	fs,
+	possibleRequests,
+	errorsAccumulator = "",
+	callback
+) {
+	const cb = (error, result) => {
+		if (error) {
+			errorsAccumulator += `${error.message}\n\n`;
+
+			const tailPossibleRequests = possibleRequests.slice(1);
+
+			if (tailPossibleRequests.length === 0) {
+				error.message = errorsAccumulator;
+
+				return callback(error);
+			}
+
+			return fetchPathsFromFilesystem(
+				fs,
+				tailPossibleRequests,
+				errorsAccumulator,
+				cb
+			);
+		}
+		callback(null, result);
+	};
+	fetchFromFilesystem(fs, possibleRequests[0], cb);
+}
+
+/**
+ * @param {InputFileSystem} fs fs
+ * @param {string} context context
+ * @param {string} url url
+ * @param {boolean} skipReading skip reading
+ * @param {function (Error|null, {sourceURL: string, sourceContent?: string}=): void} callback callback
+ * @returns {void}
+ */
+function fetchFromURL(fs, context, url, skipReading, callback) {
+	// 1. It's an absolute url and it is not `windows` path like `C:\dir\file`
+	if (/^[a-z][a-z0-9+.-]*:/i.test(url) && !path.win32.isAbsolute(url)) {
+		const { protocol } = new urlUtils.URL(url);
+
+		if (protocol === "data:") {
+			if (skipReading) return callback(null, { sourceURL: "" });
+			return fetchFromDataURL(url, (err, sourceContent) => {
+				if (err) return callback(err);
+				callback(null, { sourceURL: "", sourceContent });
+			});
+		} else if (protocol === "file:") {
+			const sourceURL = urlUtils.fileURLToPath(url);
+			if (skipReading) return callback(null, { sourceURL });
+			return fetchFromFilesystem(fs, sourceURL, (err, result) => {
+				if (err) return callback(err);
+				const { data: sourceContent } = result;
+				callback(null, { sourceURL, sourceContent });
+			});
+		} else {
+			return callback(
+				new Error(`Failed to parse source map: '${url}' URL is not supported`)
+			);
+		}
+	}
+
+	// 2. It's a scheme-relative
+	if (/^\/\//.test(url)) {
```

Comment:
Stateless: Let's hoist this and name it something obvious.

Reply:
Done.